crates/iroha_tree/src/lib.rs

```diff
@@ -0,0 +1,633 @@
+//! A transitional crate that might eventually be merged into other crates.
+//!
+//! It aims to integrate executables, events, and event filters while enabling recursive trigger prediction in a static manner.
+//! The prediction is based on the union of possible execution paths, enabling pessimistic event loop detection.
+//!
+//! Additionally, to maintain performance while enabling per-transaction triggers (#4937), it consolidates:
+//!
+//! - Instructions into a single [`ChangeSet`] per transaction.
+//! - (Data) events into a single [`Event`] per transaction.
+//! - (Data) event filters into a single [`Receptor`] per trigger.
+//! - Permissions, roles, and ownerships into a single [`Permission`] per validation.
+
+#![expect(missing_copy_implementations)]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use core::{
+    convert::Infallible,
+    fmt::{self, Debug, Display},
+    ops::{Add, BitOr},
+    str::FromStr,
+};
+
+#[cfg(not(feature = "std"))]
+extern crate alloc;
+#[cfg(not(feature = "std"))]
+use alloc::{
+    boxed::Box,
+    collections::{btree_map, BTreeMap, BTreeSet},
+    rc::Rc,
+    vec,
+    vec::Vec,
+};
+#[cfg(feature = "std")]
+use std::{
+    collections::{btree_map, BTreeMap, BTreeSet},
+    rc::Rc,
+};
+
+use derive_more::{BitOr, Constructor, DebugCustom, From};
+use parity_scale_codec::{Decode, Encode};
+use serde_with::{DeserializeFromStr, SerializeDisplay};
+
+pub mod changeset;
+pub mod event;
+pub mod permission;
+pub mod readset;
+pub mod receptor;
+pub mod state;
+
+/// A flattened node map with a fixed skeleton equivalent to the world state.
+/// Node values may vary by [`Mode`].
+#[derive(Debug, PartialEq, Eq, Clone, Decode, Encode)]
+pub struct Tree<M: Mode>(BTreeMap<NodeKey, NodeValue<M>>);
```

Comment:
If #5358 is actually planned, I'm not sure whether such a structure with heap allocation can be passed to and returned from Wasm @mversic